_drafts/schematic_review_checklist_for_firmware_engineer.md

```diff
@@ -0,0 +1,132 @@
+
+Unfortunately, devices stuck in this unresponsive mode were also in a low power state. Luckily, our low power mode was not that good at the time, but we ended up having to instruct our users to let bikes set for 4 days before they could recover. That's a pretty poor user experience.
+
+While we solved this in future releases by adding an NFC based reset, this is an issue that can only be fixed going forward, and not in reverse. Adding a physical way to reset the device for a human user from the beginning is a very powerful solution. How many things are fixed by "turning it off and turning it back on again"? 
+
+## Story 3 - I2C Resets
+While developing a fitness tracker, I kept running into a bug where I occasionally needed to fully power cycle a device to get the temperature sensor to start working again. The temperature sensor was connected over I2C, and digging into the issue I found that, when this was happening, the I2C SDA line was stuck low. We were able to reproduce that failure by having the MCU reset in the middle of an I2C read. Because the I2C controller on the MCU was controlling the clock, and stopped clocking, the peripheral was still driving the SDA line, waiting for the next clock signal.
+
+We had no reset line headed to the I2C device, so we could not reset the I2C device. However, we could manually clock out the remaining SCL cycles on the SCL line, and so modified our boot sequence to manually toggle the SCL line to "clock out" the remaining data, freeing up the I2C bus.
+
+However, I'd still have preferred a true reset line.
+
+## Story 4: GPIO Maps
+This is less of a war story, and more of a shoutout to good electrical engineers. Before walking into the schematic review for one project I was working on, the EE on the project published a spreadsheet of all the GPIOs for the MCU. This included their main function after reset, intended function, and any additional notes. It made it very easy as a firmware engineer to scan the list of GPIOs and ensure they were all configured properly. I'd recommend this is something that all firmware engineers ask for when looking at a schematic.
+
+## Story 5: Too much or too little debug
+I worked on a project once with an electrical engineer that proposed we make it hard to attach a debugger to the board. He proposed that we force someone to solder wires/leads to debug pads in order to get a debugger to work. As a young firmware engineer, I was floored. The debugger was my lifeline, and especially on a Cortex-M4 could help troubleshoot so many things. Making it difficult for a firmware engineer to use their most powerful tool seemed ludicras to me.
```

Comment:
Out of curiosity, was his concern at all related to tampering, so he wanted to make it "hard" for them? (of course they would probably just solder it themselves if they got as far as breaking it open). Or just make it hard for you 😄

Reply:
I don't think so... I don't want to speak for this other engineer, but my thought is, (as you stated) if you're reverse engineering a product, attaching leads to JTAG is likely a small step.